### DISCUSSION TITLE
how do i have a autoincrement primay key with sqlite ?

### DISCUSSION THREAD

**Discussion by w-Bro** [sqlalchemy/sqlalchemy#7306] [Usage Questions]
I have known that `sqlite_autoincrement` in doc, bui it is not work well for me, any example code for me.
The doc url is `https://docs.sqlalchemy.org/en/14/dialects/sqlite.html`

**Comment by gordthompson** [comment]
SQLAlchemy will automatically create an autoincrement column if it is an integer column that is the only column in the primary key:

https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.Column.params.autoincrement

```python
t = Table(
    "t",
    MetaData(),
    Column("id", Integer, primary_key=True),
    Column("txt", String),
)
t.create(engine)
with engine.begin() as conn:
    conn.execute(t.insert(), {"txt": "Hello world!"})
    print(conn.execute(select(t)).mappings().fetchall())
    # [{'id': 1, 'txt': 'Hello world!'}]
```

**Reply by zzzeek** [reply]
they might be asking about this:

https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#using-the-autoincrement-keyword

can you clarify what's not working?

**Reply by w-Bro** [reply]
> they might be asking about this:
> 
> https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#using-the-autoincrement-keyword
> 
> can you clarify what's not working?

Thanks a lot, I alway define table as a way of `Class`, and it's embrassing that it works well after I try again with my code.
Here is my example code, for anyone trouble in it, I have tested it in both sqlite and mysql.

base_class.py
```py
from sqlalchemy.ext.declarative import as_declarative, declared_attr
from sqlalchemy.dialects import sqlite

BIGINT = BIGINT().with_variant(sqlite.INTEGER, "sqlite")

@as_declarative()
class Base:
    # common keys
    id = Column(BIGINT, primary_key=True)
    create_time = Column(DateTime, default=datetime.now, server_default=func.now(), comment="xxx")
    update_time = Column(DateTime, default=datetime.now, onupdate=datetime.now, server_default=func.now(),
                         server_onupdate=func.now(), comment="xxx")
    __name__: str

    # Generate __tablename__ automatically
    @declared_attr
    def __tablename__(cls) -> str:
        import re
        name_list = re.findall(r"[A-Z][a-z\d]*", cls.__name__)
        return "_".join(name_list).lower()

    __table_args__ = {"sqlite_autoincrement": True}
```

test.py
```py
class User(Base):

    username = Column(VARCHAR(20))

    __tablename__ = "user"

if __name__ == '__main__':
    create_tables()
    with create_session(commit=True) as sess:
        user = User(username="!23")
        sess.add(user)
        sess.commit()
```
In my situation, i have to make sure that all db models work well in both sqlite and other relation dbs, at least mysql, and create a common base model for all of models. If the code is incorrect or any problem in it, fell free let me know it!

**Reply by zzzeek** [reply]
i dont think you need the sqlite_autoincrement feature for that.  the name of SQLite "autoincrement" is misleading because a table that has a single integer primary key column will be effectively "autoincrement" in any case, as it exposes the row id in this case (This is all sqlite-specific stuff).  "autoincrement" just makes it use a different and more expensive algorithm for calculating the rowid.

see the first line at https://www.sqlite.org/autoinc.html:

>  The AUTOINCREMENT keyword imposes extra CPU, memory, disk space, and disk I/O overhead and should be avoided if not strictly needed. It is usually not needed. 

I'd take it out.

**Reply by w-Bro** [reply]
> i dont think you need the sqlite_autoincrement feature for that. the name of SQLite "autoincrement" is misleading because a table that has a single integer primary key column will be effectively "autoincrement" in any case, as it exposes the row id in this case (This is all sqlite-specific stuff). "autoincrement" just makes it use a different and more expensive algorithm for calculating the rowid.
> 
> see the first line at https://www.sqlite.org/autoinc.html:
> 
> > The AUTOINCREMENT keyword imposes extra CPU, memory, disk space, and disk I/O overhead and should be avoided if not strictly needed. It is usually not needed.
> 
> I'd take it out.

Get new knowledge! But just like what i say, I want it work on both sqlite, mysql and other db, it's easy for us to use same code as a base project.
